### DISCUSSION TITLE
Dynamic Capacity Demands

### DISCUSSION THREAD

**Discussion by satyenataagyaa** [google/or-tools#3327] [Routing (and legacy CP) questions] [answered]
I am trying to solve a Capacity Constraints VRP, with a twist.

The problem is to pick up passengers from around town, and bring them to an office location (the depot). As such, the "demand" associated with each location is "1". The demand becomes "2", however, when the passenger being picked up is female AND is the first pickup on any route. I am assuming I need to go beyond the Transit Callback mechanism since that does not provide a view into the state of the solver.

I am new to OR-Tools, and would appreciate any pointers on how to model this problem. Thanks!

**Comment by jmarca** [comment, marked as answer]
I think one issue is that you are not being clear with your
definitions.

"Demand" means nothing in and of itself.  In the OR Tools routing
solver, there are dimensions that can add and subtract values.  The
loose logical concept of "demand" is translated into "add
one" or "subtract one" from a dimension.

If, as you say, a particular thing means "add two", then add two.  So
if a pickup node has some trait "female", then in the callback
function for that node, check for that trait and add two.

More difficult is the part where the pickup is first, but it isn't
*that* difficult.  One way to do it is to split the nodes of interest into
two nodes, only one of which can be served (use a disjunction over the
nodes with a cardinality of 1).  Then the node with value of 2 can
only be served if the vehicle arrives with a load of 0, and the node
with a value of 1 can only be served if the vehicle arrives with a
load greater than 0.

If you are sure that each vehicle will start with zero from the depot
node only, then you might be able to use that fact, but in the more
general case of passengers getting dropped anywhere, that is not so helpful.

(You need to do these gymnastics because the callback cannot know the
state of the vehicle when it arrives.   But by setting up constraints,
you can know a priori the state of the vehicle.)

I'm sure that won't solve your problem, but struggle a bit and post some
working (or lightly broken) code and someone can help more.

Finally, if your vehicles are capacity limited, one consequence of
this rule might be that no female will get picked up first...the
vehicle can fit more pax if females are not picked up first.

James

On Thu, Jun 02, 2022 at 02:42:34PM -0700, satyenataagyaa wrote:
> I am trying to solve a Capacity Constraints VRP, with a twist.
>
> The problem is to pick up passengers from around town, and bring them to an office location (the depot). As such, the "demand" associated with each location is "1". The demand becomes "2", however, when the passenger being picked up is female AND is the first pickup on any route. I am assuming I need to go beyond the Transit Callback mechanism since that does not provide a view into the state of the solver.
>
> I am new to OR-Tools, and would appreciate any pointers on how to model this problem. Thanks!
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3327
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by satyenataagyaa** [reply]
James, I appreciate the quick response. I was looking for a direction, and you have provided it with adequate explanation.

I will give your ideas a try, and post back here how I fare.

Thanks again!

**Reply by satyenataagyaa** [reply]
Splitting the nodes of interest into two nodes helped me arrive at a solution for my problem. Thank you for the suggestion.

I am seeing good results, although I have yet to try with samples larger than 50 passengers.

**Reply by satyenataagyaa** [reply]
Hi James,

I am now working on Part 2 of this problem. This time the task is to transport passengers from the depot to locations around town. The additional constraint is the need for an escort if the last passenger to be dropped is of female gender.

As with the pickup problem (above), I am trying to address this using a Capacity dimension. Using the "splitting the nodes of interest into two nodes" approach, I am using this code for vehicles with a max capacity of 4:

                # First node (demand == 1) can be used only when fewer than 3 passengers dropped already
                solver.Add(capacity_dimension.CumulVar(node_idx1) < 3)
                # Second node (demand == 2) can be used only when passengers dropped already >= 3
                solver.Add(capacity_dimension.CumulVar(node_idx2) >= 3)

This works, but I would like to get away from the hard-coded value of '3' in the code above. Also, this only works when all vehicles have the same capacity, i.e. 4. I would like the code to be able to handle a fleet with more than one capacity.

Your advice is greatly appreciated.

Satyen

**Comment by jmarca** [comment]
Yep that is tricky, and not something that I’m going to figure out for you in my spare time…

James
